### PR TITLE
Wallaby Epics

### PR DESCRIPTION
**Pull request by mephenor**

### CODE REVIEW

File: 39-red-necked-wallaby/technical_specification.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+- Keep the existing ingest and file upload commands, but rename them and mark them as legacy/deprecated (e.g. legacy-files upload/legacy-files batch-upload)
+
+- Add new upload commands to that are based on the existing ones, but have one slight difference: The output metadata contains the secret ID instead of the actual file encryption/decryption secret. This means the following changes are necessary
+    - Before writing the output metadata file, the file secret is stored and a secret ID returned by calling the new `POST /federated/ingest_secret` endpoint
+    - If the secret deposition fails, the uploaded file needs to be cleaned up, i.e. deleted after retry logic is exhausted
+
+#### File Ingest Service Changes:
+
+- Keep the existing `POST /ingest` endpoint to handle metadata for already update files
```
Comment (TheByronHimes) [inline 39-red-necked-wallaby/technical_specification.md:23]:
```suggestion
- Keep the existing `POST /ingest` endpoint to handle metadata for already updated files
```

File: 39-red-necked-wallaby/technical_specification.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+- Add two new endpoints, that split the existing functionality between them:
```
Comment (TheByronHimes) [inline 39-red-necked-wallaby/technical_specification.md:24]:
```suggestion
- Add two new endpoints that split the existing functionality between them:
```

File: 39-red-necked-wallaby/technical_specification.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    - `POST /federated/ingest_metadata` that should be used by a central data steward to ingest the metadata into the file services, i.e. this endpoint no longer communicates with the vault but directly gets the secret ID in the received file metadata
+    - `POST /federated/ingest_secret` that should be used by local data stewards during file uploads to ingest the secret into vault and return the corresponding secret ID
+
+- All endpoints should use the security measures employed by the existing endpoint, i.e. token/token hash pairs for authentication, asymmetric payload encryption using Crypt4GH keypairs and communication only over HTTPS
+
+#### GHGA Connector Changes:
+
+- The Connector currently is a bit brittle with regard to being modifyable/extendable. Existing upload/download code should be refactored to a more object oriented approach, using abstract base classes to decouple concerns between different parts of the functionality where possible.
```
Comment (TheByronHimes) [inline 39-red-necked-wallaby/technical_specification.md:32]:
```suggestion
- The Connector currently is a bit brittle with regard to being modifiable/extendable. Existing upload/download code should be refactored to a more object oriented approach, using abstract base classes to decouple concerns between different parts of the functionality where possible.
```

File: 39-red-necked-wallaby/technical_specification.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+#### Upload Controller Changes:
+
+- The upload controller is currently the most outdated service in the file backend microservice landscape. Thus two subtasks are needed to bring it up to state with the other microservices: 1) Investigate, what needs to be changed/updated and 2) implement the required changes.
```
Comment (TheByronHimes) [inline 39-red-necked-wallaby/technical_specification.md:36]:
```suggestion
- The upload controller is currently the most outdated service in the file backend microservice landscape. Thus two subtasks are needed to bring it up to state with the other microservices: 1) Investigate what needs to be changed/updated and 2) implement the required changes.
```

**Review by TheByronHimes — DISMISSED** [review]
Only nitpicky suggestions. 
One possible change however:
The term 'object storage location' here (in the epic, not just this paragraph) is a little confusing because I immediately interpret it as 'the bucket the file is in'. But you're referring the physical site location, such as DKFZ/Tübingen, right? Something less ambiguous/overloaded or would be better, like data hub, GHGA node or something along those lines.

File: 40-agile-wallaby/technical_specification.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+Moving from separate `access_key_id`, `secret_access_key` and `endpoint_url` for one storage configuration, a `s3_object_storages` object will now hold the configuration for all available locations.
+Each location is identified by a label and the nested dict holds both the bucket name that shall be accessed as well as the credentials and endpoint URL for the specified bucket at the given location.
+
+```
+s3_object_storages:
+    DKFZ:
+        permanent_bucket: ghga_permanent
+        credentials:
+            s3_access_key_id: test
+            s3_endpoint_url: http://localstack:4566
```
Comment (KerstenBreuer) [inline 40-agile-wallaby/technical_specification.md:25]: Maybe move the endpoint URL out of credentials.

File: 40-agile-wallaby/technical_specification.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            s3_secret_access_key: '**********'
+            s3_session_token: null
```
Comment (KerstenBreuer) [inline 40-agile-wallaby/technical_specification.md:27]: The "s3_" prefix can be omitted here.

**Review by KerstenBreuer — DISMISSED** [review]
Well written, just small suggestions

File: 40-agile-wallaby/technical_specification.md (hunk 5 above)
Comment (mephenor) [reply 40-agile-wallaby/technical_specification.md:25]: Would probably make more sense to rename credentials instead, so we can reuse the existing S3Config from hexkit

Comment (KerstenBreuer) [reply 40-agile-wallaby/technical_specification.md:25]: Sounds good as well

**Review by KerstenBreuer — APPROVED** [review]

**Review by TheByronHimes — APPROVED** [review]

**Pull request merged**